src/client/chat.js

```diff
@@ -0,0 +1,538 @@
+        }, options.timeoutPreview)
+      }
+    }
+
+    if (diff < 100) {
+      setTimeout(sendPacket, diff + 1)
+    } else sendPacket()
+  }
+
+  function queueMessagePreview (message, resolve, reject) {
+    if (!client.previewQueue) client.previewQueue = []
+    client.previewQueue.push({ message, resolve, reject })
+  }
+
+  // Packet handling
+
+  client.on('chat_preview', handleChatPreview)
+
+  client.on('player_info', handlePlayerInfo)
+
+  client.on('player_remove', handlePlayerRemove)
+
+  client.on('player_chat', handlePlayerChat)
+
+  client.on('system_chat', handleSystemChat)
+
+  client.on('hide_message', handleHideChat)
+
+  client.on('message_header', handleChatHeader)
+
+  client.on('should_display_chat_preview', packet => {
+    client.useChatPreview = packet.should_display_chat_preview
+  })
+
+  client.on('server_data', packet => {
+    client.useChatPreview = packet.previewsChat
+    client.signedChat = packet.enforcesSecureChat
+  })
+
+  function handleChatPreview (packet) {
+    if (packet.queryId === client.pendingPreview.id) {
+      setTimeout(() => {
+        client.pendingPreview.resolve(packet.message)
+      }, 200) // Previews are invalid for the first 200 milliseconds
+
+      if (client.previewTimeout) clearTimeout(client.previewTimeout)
+      if (options.queueMessages && !!client.previewQueue) {
+        const next = client.previewQueue.splice(0, 1)
+        if (next) requestPreview(next.message, next.resolve, next.reject)
+      }
+    }
+  }
+
+  function handlePlayerRemove (packet) {
+    packet.players.forEach(player => {
+      if (!client.players || !client.players[player.UUID]) return
+      delete client.players[player.UUID]
+    })
+  }
+
+  function handlePlayerInfo (packet) {
+    if (mcData.supportFeature('playerInfoActionsIsBitfield')) { // 1.19.3
+      if (packet.action & 2) { // chat session
+        if (!client.players) client.players = {}
+        packet.data.forEach(player => {
+          if (!player.chatSession) return
+          client.players[player.UUID] = {
+            uuid: player.chatSession.uuid,
+            publicKey: player.chatSession.publicKey.keyBytes
+          }
+        })
+      }
+    } else { // 1.19.2 and earlier
+      if (packet.action === 0) { // add player
+        if (!client.players) client.players = {}
+        packet.data.forEach(player => {
+          client.players[player.UUID] = player.crypto
+        })
+      } else if (packet.action === 4) { // remove player
+        packet.data.forEach(player => {
+          if (!client.players || !client.players[player.UUID]) return
+          delete client.players[player.UUID]
+        })
+      }
+    }
+  }
+
+  function handleHideChat (packet) {
+    if (mcData.supportFeature('acknowledgeUntracked')) {
+      const signature = packet.signature || resolveSignature(packet.id)
+      if (signature) removeMessage(packet.signature)
```

Comment:
I don't think it's removed from the lastSeen messages array, in 1.19.2 it appears to run [`removeFromDelayedMessageQueue`](https://github.dev/extremeheat/extracted_minecraft_data/blob/client1.19.1/client/net/minecraft/client/multiplayer/ClientPacketListener.java#L783) and only removes from inbound message throttle

Reply:
This code only runs for 1.19.3. That's what the feature check is for

Reply:
Is this the correct behavior on .3?

Reply:
Yes

Reply:
This code doesn't run on 1.19.2 clients right ?

Reply:
Correct